packages/grid/_modules_/grid/hooks/features/useComponents.tsx

```diff
@@ -1,93 +1,73 @@
+import {
+  BaseComponentProps,
+  ApiRef,
+  GridSlotsComponent,
+  RootContainerRef,
+  DEFAULT_SLOTS_COMPONENTS,
+} from '../../models';
+import { ErrorOverlayProps } from '../../components/ErrorOverlay';
+import { optionsSelector } from '../utils/optionsSelector';
 import { visibleColumnsSelector } from './columns/columnsSelector';
 import { useGridSelector } from './core/useGridSelector';
-import { paginationSelector } from './pagination/paginationSelector';
+import { useGridState } from './core/useGridState';
 import { unorderedRowModelsSelector } from './rows/rowsSelector';
 
+function EmptyComponent() {
+  return null;
+}
+
 export const useComponents = (
-  componentOverrides: GridComponentOverridesProp | undefined,
+  componentsProp: GridSlotsComponent | undefined,
   apiRef: ApiRef,
   gridRootRef: RootContainerRef,
 ) => {
   const options = useGridSelector(apiRef, optionsSelector);
   const rows = useGridSelector(apiRef, unorderedRowModelsSelector);
   const columns = useGridSelector(apiRef, visibleColumnsSelector);
-  const pagination = useGridSelector(apiRef, paginationSelector);
+  const [state] = useGridState(apiRef!);
 
-  const componentProps: ComponentProps = React.useMemo(
+  const baseComponentProps: BaseComponentProps = React.useMemo(
     () => ({
-      pagination,
+      state,
       rows,
       columns,
       options,
       api: apiRef,
       rootElement: gridRootRef,
     }),
-    [pagination, rows, columns, options, apiRef, gridRootRef],
-  );
-
-  const headerComponent = React.useMemo(
-    () =>
-      componentOverrides?.header
-        ? React.createElement(componentOverrides.header, componentProps)
-        : null,
-    [componentOverrides, componentProps],
+    [state, rows, columns, options, apiRef, gridRootRef],
   );
 
-  const footerComponent = React.useMemo(
-    () =>
-      componentOverrides?.footer
-        ? React.createElement(componentOverrides.footer, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
+  const components = React.useMemo(() => {
+    const allComponents = { ...DEFAULT_SLOTS_COMPONENTS, ...componentsProp };
 
-  const loadingComponent = React.useMemo(
-    () =>
-      componentOverrides?.loadingOverlay ? (
-        React.createElement(componentOverrides.loadingOverlay, componentProps)
-      ) : (
-        <LoadingOverlay />
-      ),
-    [componentOverrides, componentProps],
-  );
-  const noRowsComponent = React.useMemo(
-    () =>
-      componentOverrides?.noRowsOverlay ? (
-        React.createElement(componentOverrides.noRowsOverlay, componentProps)
-      ) : (
-        <NoRowMessage />
-      ),
-    [componentOverrides, componentProps],
-  );
+    const wrapWithBaseProps: <TProps>(
+      Component: React.ElementType<TProps & BaseComponentProps> | undefined | null,
+    ) => React.ElementType<TProps> = <TProps extends {}>(Component) => {
+      if (Component == null) {
+        return EmptyComponent;
+      }
```

Comment:
This seems to yield a different behavior to the other components. From what I understand, if we do:

```jsx
<DataGrid components={{ Footer: undefined }} />
```

It's currently equivalent to:

```jsx
<DataGrid components={{ Footer: () => null }} />
```

while if we follow the other component, it should be equivalent to:

```jsx
<DataGrid components={{ Footer: GridFooter }} />
```

cc @mnajdova